numpy/_core/tests/test_arrayobject.py

```diff
@@ -31,3 +31,39 @@ def test_matrix_transpose_equals_swapaxes(shape):
     tgt = np.swapaxes(arr, num_of_axes - 2, num_of_axes - 1)
     mT = arr.mT
     assert_array_equal(tgt, mT)
+
+
+@pytest.mark.parametrize("subclass", [False, True])
+def test_array_wrap(subclass):
+    # NumPy should allow `__array_wrap__` to be called on arrays, it's logic
+    # is designed in a way that:
+    #
+    # * Subclasses never return scalars by default (to preserve their
+    #   information).  They can choose to if they wish.
+    # * NumPy returns scalars, if `return_scalar` is passed as True to allow
+    #   manual calls to `arr.__array_wrap__` to do the right thing.
+
+    class MyArr(np.ndarray):
+        def __array_wrap__(self, arr, context=None, return_scalar=None):
+            return super().__array_wrap__(arr, context, return_scalar)
+
+    arr = np.arange(3)
+    if subclass:
+        arr = arr.view(MyArr)
+
+    arr0d = np.array(3, dtype=np.int8)
```

Comment:
What if `arr0d` is a subclass too? As written, I think that causes decay to a scalar (if `arr` is a subclass).

Reply:
True, but I don't think it should ever be a subclass!  I.e. that would defeat the purpose of wrapping (converting everything to plain arrays)?
(Technically, subclasses are accepted here.  But in practice, I think we must assume that the original subclass is meaningless, because we must return *our* subclass.)

Reply:
Ah, sorry, I misunderstood.  While it is probably true that this shouldn't happen, I agree that I should anticipate it and fix the branch (return of scalar must only happens if `self` is a base-class array)